zero/gkyl_position_map_priv.h

```diff
@@ -0,0 +1,448 @@
+#include <gkyl_array.h>
+#include <gkyl_calc_bmag.h>
+
+struct opt_Theta_ctx
+{
+  struct gkyl_position_map *gpm;
+  struct gkyl_bmag_ctx *bmag_ctx;
+  double dB_target;
+  double dB_global_lower;
+  double B_lower_region;
+};
+
+// Allocate double array (filled with zeros).
+static struct gkyl_array*
+mkarr(bool on_gpu, long nc, long size)
+{
+  struct gkyl_array* a;
+  a = gkyl_array_new(GKYL_DOUBLE, nc, size);
+  return a;
+}
+
+/**
+ * Function that actually frees memory associated with this
+ * object when the number of references has decreased to zero.
+ *
+ * @param ref Reference counter for this object.
+ */
+void gkyl_position_map_free(const struct gkyl_ref_count *ref);
+
+// Utility functions for polynomial based B mapping
+
+void
+calculate_mirror_throat_location_polynomial(struct gkyl_position_map_const_B_ctx *constB_ctx, struct gkyl_bmag_ctx *bmag_ctx)
+{
+  // Assumes symmetry along theta centered at 0 and two local maxima in Bmag that are symmetric
+
+  enum { X_IDX, Y_IDX, Z_IDX }; // arrangement of cartesian coordinates
+  double psi = constB_ctx->psi;
+  double alpha = constB_ctx->alpha;
+  double *xp = malloc(3*sizeof(double));
+  xp[X_IDX] = psi;
+  xp[Y_IDX] = alpha;
+  int itterations = 10;
+  double interval_left = 0.0;
+  double interval_right = constB_ctx->theta_max;
+  int points_per_level = 20;
+  double maximum_Bmag = 0.0;
+  double maximum_Bmag_location = 0.0;
+  double *fout = malloc(3*sizeof(double));
+  for (int j = 0; j < itterations; j++)
+  {
+    double dz = (interval_right - interval_left) / points_per_level;
+    maximum_Bmag = 0.0;
+    maximum_Bmag_location = 0.0;
+    for (int i = 0; i < points_per_level; i++)
+    {
+      double z = interval_left + i * dz;
+      xp[Z_IDX] = z;
+      gkyl_calc_bmag_global(0.0, xp, fout, bmag_ctx);
+      double Bmag = fout[0];
+      if (Bmag > maximum_Bmag)
+      {
+        maximum_Bmag = Bmag;
+        maximum_Bmag_location = z;
+      }
+    }
+    interval_left = maximum_Bmag_location - dz;
+    interval_right = maximum_Bmag_location + dz;
+  }
+  constB_ctx->theta_throat = maximum_Bmag_location;
+  constB_ctx->Bmag_throat = maximum_Bmag;
+  free(xp);
+  free(fout);
+}
+
+void
+position_map_constB_z_polynomial(double t, const double *xn, double *fout, void *ctx)
+{
+  // Converts our uniform coordinate along field line length to a non-uniform coordinate
+  // according to the polynomial mapping of arbitrary order.
+  // Notation: I switch from theta to z here. Both are computational coordinate.
```

Comment:
The reader doesn't know who "I" is. Either speak in 3rd person or initialize and date your comment, e.g. MR 2024/12/25.